<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Async - Await</title>
</head>
<body>

    <h1>Async - Await</h1>

    <script>

        function cuadradoPromise(value){

        if(typeof value !== "number") return Promise.reject(`Èrror, el valor ${value} no es un número`);

        return new Promise((resolve, reject)=>{

            setTimeout(() => {

                resolve(

                    {
                    value,
                    result: value*value
                    }
                
                )
                
            }, 0 | Math.random()*100)

        })

        }

        async function funcionAsyncDeclarada(){

            try {

                console.log("Inicio de AsyncFunction")
                let obj = await cuadradoPromise(0)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)

                obj = await cuadradoPromise(1)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)
                
                obj = await cuadradoPromise(2)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)
                
                obj = await cuadradoPromise(3)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)
                
                obj = await cuadradoPromise(4)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)
                
                obj = await cuadradoPromise(5)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)
                
                obj = await cuadradoPromise(6)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)

                console.log("Fin AsyncFunction")

            } catch (error) {
                
                console.log(error)

            }

        }

        funcionAsyncDeclarada()

        const funcionAsyncExpresada = async () => {

            try {

                console.log("Inicio de AsyncFunction")
                let obj = await cuadradoPromise(7)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)

                obj = await cuadradoPromise(8)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)
                
                obj = await cuadradoPromise(9)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)
                
                obj = await cuadradoPromise(10)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)
                
                obj = await cuadradoPromise(11)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)
                
                obj = await cuadradoPromise(12)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)
                
                obj = await cuadradoPromise(13)
                console.log(`AsyncFunction: ${obj.value}, ${obj.result}`)

                console.log("Fin AsyncFunction")

            } catch (error) {
                
                console.log(error)

            }

        }

        funcionAsyncExpresada()

    </script>
    
</body>
</html>